packages/block-library/src/navigation/index.php

```diff
@@ -67,6 +67,521 @@ function block_core_navigation_sort_menu_items_by_parent_id( $menu_items ) {
+		$class = WP_Navigation_Block_Renderer::get_classes( $attributes );
+		$container_attributes = get_block_wrapper_attributes(
+			array(
+				'class' => 'wp-block-navigation__container ' . $class,
+				'style' => $style,
+			)
+		);
+
+		$inner_blocks_html = '';
+		$is_list_open      = false;
+
+		foreach ( $inner_blocks as $inner_block ) {
+			$is_list_item = in_array( $inner_block->name, $list_item_nav_blocks, true );
+
+			if ( $is_list_item && ! $is_list_open ) {
+				$is_list_open       = true;
+				$inner_blocks_html .= sprintf(
+					'<ul %1$s>',
+					$container_attributes
+				);
+			}
+
+			if ( ! $is_list_item && $is_list_open ) {
+				$is_list_open       = false;
+				$inner_blocks_html .= '</ul>';
+			}
+
+			$inner_block_content = $inner_block->render();
+			if ( ! empty( $inner_block_content ) ) {
+				if ( in_array( $inner_block->name, $needs_list_item_wrapper, true ) ) {
+					$inner_blocks_html .= '<li class="wp-block-navigation-item">' . $inner_block_content . '</li>';
+				} else {
+					$inner_blocks_html .= $inner_block_content;
+				}
+			}
+		}
+
+		if ( $is_list_open ) {
+			$inner_blocks_html .= '</ul>';
+		}
+
+		// Add directives to the submenu if needed.
+		if ( $has_submenus && $should_load_view_script ) {
+			$w                 = new WP_HTML_Tag_Processor( $inner_blocks_html );
+			$inner_blocks_html = block_core_navigation_add_directives_to_submenu( $w, $attributes );
+		}
+
+		return $inner_blocks_html;
+	}
+
+	/**
+	 * Gets the inner blocks for the navigation block.
+	 */
+	private static function get_inner_blocks_for_navigation( $block, $attributes ) {
+		$inner_blocks = $block->inner_blocks;
+
+		// Ensure that blocks saved with the legacy ref attribute name (navigationMenuId) continue to render.
+		if ( array_key_exists( 'navigationMenuId', $attributes ) ) {
+			$attributes['ref'] = $attributes['navigationMenuId'];
+		}
+
+		// If:
+		// - the gutenberg plugin is active
+		// - `__unstableLocation` is defined
+		// - we have menu items at the defined location
+		// - we don't have a relationship to a `wp_navigation` Post (via `ref`).
+		// ...then create inner blocks from the classic menu assigned to that location.
+		if (
+			defined( 'IS_GUTENBERG_PLUGIN' ) && IS_GUTENBERG_PLUGIN &&
+			array_key_exists( '__unstableLocation', $attributes ) &&
+			! array_key_exists( 'ref', $attributes ) &&
+			! empty( block_core_navigation_get_menu_items_at_location( $attributes['__unstableLocation'] ) )
+		) {
+			$menu_items = block_core_navigation_get_menu_items_at_location( $attributes['__unstableLocation'] );
+			if ( empty( $menu_items ) ) {
+				return '';
+			}
+
+			$menu_items_by_parent_id = block_core_navigation_sort_menu_items_by_parent_id( $menu_items );
+			$parsed_blocks           = block_core_navigation_parse_blocks_from_menu_items( $menu_items_by_parent_id[0], $menu_items_by_parent_id );
+			$inner_blocks            = new WP_Block_List( $parsed_blocks, $attributes );
+		}
```

Comment:
We should break all conditional branches into their own methods.

- for_unstable_location
- from_navigation post
- from_fallback

Reply:
Done in [cd2c7fd](https://github.com/WordPress/gutenberg/pull/55605/commits/cd2c7fd5fe441297f3aa8a753b089091d376088e)